File: examples/typescript/fetch-mot-history.ts
// examples/typescript/fetch-mot-history.ts
//
// Minimal example: pay for and fetch a single MOT history lookup via x402.
//
// Setup:
//   npm install @x402/fetch @x402/evm viem
//   export AGENT_PRIVATE_KEY=0x...   (a Base mainnet wallet funded with a small amount of USDC)
//
// Run:
//   npx tsx fetch-mot-history.ts AB12CDE

import { wrapFetchWithPayment, x402Client } from '@x402/fetch';
import { registerExactEvmScheme } from '@x402/evm/exact/client';
import { privateKeyToAccount } from 'viem/accounts';

const BASE_URL = 'https://api.databroker.mossforge.dev';

interface DataBrokerError {
    error: {
        code: string;
        message: string;
        retryable: boolean;
        retry_after_seconds?: number;
        docs_url: string;
    };
}

interface MotHistoryResponse {
    dataset: string;
    key: string;
    fetched_at: string;
    fetch_status: 'ok' | 'error' | 'partial';
    source: string;
    ttl_seconds: number;
    stale: boolean;
    data: {
        vehicle: {
            registration: string;
            make: string | null;
            model: string | null;
            primaryColour: string | null;
            fuelType: string | null;
            engineSize: string | null;
            firstUsedDate: string | null;
            registrationDate: string | null;
            manufactureDate: string | null;
        };
        summary: {
            latestTestDate: string | null;
            latestTestResult: 'PASSED' | 'FAILED' | null;
            latestExpiryDate: string | null;
            totalTests: number;
            passCount: number;
            failCount: number;
            passRate: number | null;
            mileageAtLatestTest: number | null;
            mileageUnit: string | null;
        };
        motTests: unknown[];
    };
}

function sleep(ms: number): Promise<void> {
    return new Promise(resolve => setTimeout(resolve, ms));
}

async function getMotHistory(
    paidFetch: typeof fetch,
    registration: string,
): Promise<MotHistoryResponse> {
    const response = await paidFetch(`${BASE_URL}/v1/dvsa-mot/${registration}`);

    if (response.ok) {
        return response.json() as Promise<MotHistoryResponse>;
    }

    const body = await response.json() as DataBrokerError;
    const { code, message, retryable, retry_after_seconds } = body.error;

    // First-ever lookup for a registration triggers a live DVSA fetch.
    // If it didn't complete within the request window, the server asks us
    // to retry — this is the one case worth handling automatically here.
    if (code === 'INGEST_IN_PROGRESS' && retryable) {
        const waitMs = (retry_after_seconds ?? 5) * 1000;
        console.log(`Data still being fetched, retrying in ${waitMs}ms...`);
        await sleep(waitMs);
        return getMotHistory(paidFetch, registration);
    }

    throw new Error(`DataBroker error [${code}]: ${message}`);
}

async function main() {
    const registration = process.argv[2];
    if (!registration) {
        console.error('Usage: tsx fetch-mot-history.ts <REGISTRATION>');
        process.exit(1);
    }

    const privateKey = process.env.AGENT_PRIVATE_KEY;
    if (!privateKey) {
        console.error('Set AGENT_PRIVATE_KEY to a funded Base mainnet wallet private key.');
        process.exit(1);
    }

    // Optional: check pricing before paying anything.
    const discoverResp = await fetch(`${BASE_URL}/v1/discover`);
    const { datasets } = await discoverResp.json() as { datasets: Array<{ dataset_id: string; price_usdc: string }> };
    const motPricing = datasets.find(d => d.dataset_id === 'dvsa-mot');
    console.log(`dvsa-mot costs $${motPricing?.price_usdc ?? '?'} USDC per call`);

    const account = privateKeyToAccount(privateKey as `0x${string}`);
    const x402 = new x402Client();
    registerExactEvmScheme(x402, { signer: account });
    const paidFetch = wrapFetchWithPayment(fetch, x402);

    console.log(`Fetching MOT history for ${registration}...`);
    const history = await getMotHistory(paidFetch, registration);

    console.log(JSON.stringify(history.data.summary, null, 2));
}

main().catch(err => {
    console.error(err.message ?? err);
    process.exit(1);
});

File: examples/python/fetch_mot_history.py
"""
examples/python/fetch_mot_history.py

Minimal example: pay for and fetch a single MOT history lookup via x402.

Setup:
    pip install x402 httpx eth-account
    export AGENT_PRIVATE_KEY=0x...   # a Base mainnet wallet funded with a small amount of USDC

Run:
    python fetch_mot_history.py AB12CDE
"""

import asyncio
import os
import sys

import httpx
from eth_account import Account
from x402.httpx import x402_payment_hooks

BASE_URL = "https://api.databroker.mossforge.dev"


class DataBrokerError(Exception):
    def __init__(
        self,
        code: str,
        message: str,
        retryable: bool,
        retry_after_seconds: int | None = None,
    ):
        super().__init__(f"DataBroker error [{code}]: {message}")
        self.code = code
        self.retryable = retryable
        self.retry_after_seconds = retry_after_seconds


async def get_mot_history(client: httpx.AsyncClient, registration: str) -> dict:
    response = await client.get(f"{BASE_URL}/v1/dvsa-mot/{registration}")

    if response.status_code < 400:
        return response.json()

    body = response.json()
    err = body["error"]

    # First-ever lookup for a registration triggers a live DVSA fetch. If it
    # didn't complete within the request window, the server asks us to retry —
    # this is the one case worth handling automatically here.
    if err["code"] == "INGEST_IN_PROGRESS" and err.get("retryable"):
        wait_seconds = err.get("retry_after_seconds", 5)
        print(f"Data still being fetched, retrying in {wait_seconds}s...")
        await asyncio.sleep(wait_seconds)
        return await get_mot_history(client, registration)

    raise DataBrokerError(
        code=err["code"],
        message=err["message"],
        retryable=err.get("retryable", False),
        retry_after_seconds=err.get("retry_after_seconds"),
    )


async def main() -> None:
    if len(sys.argv) < 2:
        print("Usage: python fetch_mot_history.py <REGISTRATION>")
        sys.exit(1)

    registration = sys.argv[1]

    private_key = os.environ.get("AGENT_PRIVATE_KEY")
    if not private_key:
        print("Set AGENT_PRIVATE_KEY to a funded Base mainnet wallet private key.")
        sys.exit(1)

    # Optional: check pricing before paying anything.
    async with httpx.AsyncClient() as discover_client:
        discover_resp = await discover_client.get(f"{BASE_URL}/v1/discover")
        datasets = discover_resp.json()["datasets"]
        mot_pricing = next((d for d in datasets if d["dataset_id"] == "dvsa-mot"), None)
        price = mot_pricing["price_usdc"] if mot_pricing else "?"
        print(f"dvsa-mot costs ${price} USDC per call")

    account = Account.from_key(private_key)

    async with httpx.AsyncClient(event_hooks=x402_payment_hooks(account)) as client:
        print(f"Fetching MOT history for {registration}...")
        history = await get_mot_history(client, registration)
        print(history["data"]["summary"])


if __name__ == "__main__":
    asyncio.run(main())
